src/igraph_eulerian.c

```diff
@@ -0,0 +1,502 @@
+/* -*- mode: C -*-  */
+/* vim:set ts=4 sw=4 sts=4 et: */
+/*
+   IGraph library.
+   Copyright (C) 2005-2012  Gabor Csardi <csardi.gabor@gmail.com>
+   334 Harvard street, Cambridge, MA 02139 USA
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301 USA
+
+*/
+
+#include "igraph_eulerian.h"
+#include "igraph_structural.h"
+#include "igraph_transitivity.h"
+#include "igraph_paths.h"
+#include "igraph_math.h"
+#include "igraph_memory.h"
+#include "igraph_random.h"
+#include "igraph_adjlist.h"
+#include "igraph_interface.h"
+#include "igraph_progress.h"
+#include "igraph_interrupt_internal.h"
+#include "igraph_centrality.h"
+#include "igraph_components.h"
+#include "igraph_constructors.h"
+#include "igraph_conversion.h"
+#include "igraph_types_internal.h"
+#include "igraph_dqueue.h"
+#include "igraph_attributes.h"
+#include "igraph_neighborhood.h"
+#include "igraph_topology.h"
+#include "igraph_qsort.h"
+#include "config.h"
+#include "structural_properties_internal.h"
+
+#include <assert.h>
+
+/* solution adapted from https://www.geeksforgeeks.org/eulerian-path-and-circuit/
+The function returns one of the following values
+0 -> if graph is not Eulerian
+1 -> if graph has an euler circuit
+2 -> if graph has an euler path */
+igraph_integer_t is_eulerian_undirected(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+
+    igraph_bool_t res;
+    igraph_inclist_t il;
+    igraph_vector_int_t *incedges;
+    igraph_integer_t odd, i;
+
+    odd = 0;
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res, IGRAPH_WEAK));
+    if (res == 0) {
+        *has_path = 0;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_ALL));
```

Comment:
Is there any reason you are building a full inclist just to check the degree? We have `igraph_degree`.